### PR TITLE
vim-patch:9.0.{1530,1533,1542,1543}

### LINKED ISSUE #12218: Cursor is moved down to a wrong line when &foldmethod is 'diff'
### Steps to reproduce

This is related to [this problem](https://github.com/vim/vim/issues/12211) .
When the step #4 is:

4. `:set foldmethod=diff`

The cursor is jumped from line 3 to line 6.
This problem happens if &fooldmethod is other than 'manual'.

### Expected behaviour

Cursor should be moved down to the next line.

### Version of Vim

9.0.0807 or later (no problem on 9.0.0806)

### Environment

Windows 10

### Logs and stack traces

_No response_

### PR DESCRIPTION
**Pull request by luukvbaal**
**vim-patch:9.0.1530: cursor moves to wrong line when 'foldmethod' is "diff"**

Problem:    Cursor moves to wrong line when 'foldmethod' is "diff". (Rick
            Howe)
Solution:   Adjust logic for scrolling. (Luuk van Baal, closes vim/vim#12364,
            closes vim/vim#12218)

https://github.com/vim/vim/commit/aa6ba308a1498dc8da04d1d30ec0470018bf782a

**vim-patch:9.0.1533: test for 'smoothscroll' is ineffective**

Problem:    Test for 'smoothscroll' is ineffective.
Solution:   Change the order of testing "zb" and "zt". (Luuk van Baal,
            closes vim/vim#12366)

https://github.com/vim/vim/commit/6f37e530d3e2d58ff055723047bf91d91af2632c

**vim-patch:9.0.1542: line not fully displayed if it doesn't fit in the screen**

Problem:    Line not fully displayed if it doesn't fit in the screen.
Solution:   Do not reset s_skipcol if not needed. (Luuk van Baal,
            closes vim/vim#12376)

https://github.com/vim/vim/commit/6c018680be0ec25d42614a93be1ea08df29a9e2a

**vim-patch:9.0.1543: display errors when making topline shorter**

Problem:    Display errors when making topline shorter and 'smoothscroll' is
            set.
Solution:   Reset w_skipcol when the topline becomes shorter than its current
            value. (Luuk van Baal, closes vim/vim#12367)

https://github.com/vim/vim/commit/5d01f86d99bc3a3fd92d4f4e9338a9e78e9ebe16

### CODE REVIEW

File: test/functional/legacy/scroll_opt_spec.lua [hunk 1]
```diff
@@ -780,6 +773,38 @@ describe('smoothscroll', function()
     ]])
   end)
 
+  -- oldtest: Test_normal_j_below_botline()
+  it([[disabled "j" does not skip lines when scrolling below botline and 'foldmethod' is not "manual"]], function()
```
Comment (zeertzjq) [inline test/functional/legacy/scroll_opt_spec.lua:777]: Why is there a "disabled" in this description

Comment (zeertzjq) [reply test/functional/legacy/scroll_opt_spec.lua:777]: Oh, I think it's better to move it out of this `describe()` block then.

Comment (luukvbaal) [reply test/functional/legacy/scroll_opt_spec.lua]: I moved it to a new file then

File: test/functional/legacy/scroll_opt_spec.lua [hunk 2]
```diff
@@ -589,14 +589,7 @@ describe('smoothscroll', function()
...
     feed('3Gztj')
-    screen:expect([[
-      <<<th lots of text with lots of text wit|
-      h lots of text with lots of text with lo|
-      ts of text with lots of text with lots o|
-      f text with lots of text end            |
-      ^four                                    |
-                                              |
-    ]])
+    screen:expect_unchanged()
```
Comment (luukvbaal) [inline test/functional/legacy/scroll_opt_spec.lua:592]: After changing the order of the tests before this, this now triggers an assert in screen.lua🤔:
```
test/functional/helpers.lua:244: test/functional/ui/screen.lua:656: _handle_grid_cursor_goto failed
  payload: { "grid_cursor_goto", { 1, -5, 0 } }
  error:   test/functional/ui/screen.lua:790: assertion failed!
```
https://github.com/neovim/neovim/blob/4e5061dba765df2a74ac4a8182f6e7fe21da125d/test/functional/ui/screen.lua#L790
What is the significance of that assert @bfredl? Can it be that it is harmless here?

Comment (luukvbaal) [reply test/functional/legacy/scroll_opt_spec.lua:592]: I think the problem is in `f_setline()`. A corrupt display/cursor position is visible after the `call setline()` step in that test. It is present in vim as well. i guess skipcol is not or not correctly computed after changing the buffer content through `setline()`.

Comment (zeertzjq) [reply test/functional/legacy/scroll_opt_spec.lua:592]: If you have a split above you'll see what's the problem.

Comment (luukvbaal) [reply test/functional/legacy/scroll_opt_spec.lua:592]: This fixes the problem, I'll submit a patch to vim.

```diff
@@ -272,6 +274,14 @@ set_buffer_lines(
                changed_bytes(lnum, 0);
                if (is_curbuf && lnum == curwin->w_cursor.lnum)
                    check_cursor_col();
+               // Reset "w_skipcol" if a currently visible line is replaced.
+               FOR_ALL_TAB_WINDOWS(tp, wp) {
+                   if (wp->w_buffer == buf
+                           && lnum < wp->w_botline
+                           && lnum >= wp->w_topline) {
+                       wp->w_skipcol = 0;
+                   }
+               }
                rettv->vval.v_number = 0;       // OK
            }
        }
```

**Review by zeertzjq — APPROVED** [review]

**Comment by zeertzjq** [comment]
> https://github.com/neovim/neovim/commit/6f41eaa2b5abd5c252428ba278a9fcc356e48c1d is ok.

Isn't that the last commit in this PR?

**Comment by Shougo** [comment]
OK.  I will check it.

**Comment by Shougo** [comment]
d6e898b44fbd6b18b3ca23f780ffaedc343961f1 broken
30a0299bc6d1ceedd04d897cf56b298dd3ded0cd OK

So d6e898b44fbd6b18b3ca23f780ffaedc343961f1 (This PR) is broken commit.

Of course, neovim HEAD is also broken.

**Comment by luukvbaal** [comment]
Care to share reproduction steps @Shougo? I haven't encountered any problems.

**Comment by Shougo** [comment]
If I have set `set smoothscroll` in terminal buffer, the problem seems occurred.

**Comment by luukvbaal** [comment]
> If I have set set smoothscroll in terminal buffer, the problem seems occurred.

> Execute git commands seems reproduced.

Can't really reproduce from that sorry. Anyways, please open an issue as it does sound like a serious problem. Please include minimal reproduction steps from `nvim --clean` or at least the output of `:set`. If the problem is really with `'smoothscroll'` we need other option values that affect it as I don't see the problem.

**Comment by Shougo** [comment]
I am investing the problem...

It can be reproduce-able with my config.  But it cannot reproduce-able with minimal config.

And it only reproduce-able with `setlocal smoothscroll` in terminal buffer.
So it needs more time.

**Comment by Shougo** [comment]
Other people does not see the problem.  So it depends on the config.
And it can be avoided by `smoothscroll` option.

**Pull request merged**